### PR TITLE
Add multicast support for udp.

### PR DESCRIPTION
**Pull request by boaks**
Basic support for multicast, if your environment supports it :-)..
For those, who think this is useful: please test and give us your feedback!

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

File: demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import org.eclipse.californium.core.network.config.NetworkConfig.Keys;
+import org.eclipse.californium.core.network.config.NetworkConfigDefaultHandler;
+
+public class MulticastTestClient {
+
+	/**
+	 * File name for network configuration.
+	 */
+	private static final File CONFIG_FILE = new File("CaliforniumMulticast.properties");
+	/**
```
Comment (vikram919) [inline demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java]: line spacing :)

File: demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	 * Header for network configuration.
+	 */
+
```
Comment (vikram919) [inline demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java]: also remove line here

File: demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	private static final String CONFIG_HEADER = "Californium CoAP Properties file for Multicast Client";
+	/**
```
Comment (vikram919) [inline demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java]: add new line here

File: demo-apps/cf-helloworld-server/src/main/java/org/eclipse/californium/examples/MulticastTestServer.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import org.eclipse.californium.elements.Connector;
+import org.eclipse.californium.elements.UdpMulticastConnector;
+
+public class MulticastTestServer {
+
+	/**
+	 * File name for network configuration.
+	 */
+	private static final File CONFIG_FILE = new File("CaliforniumMulticast.properties");
+	/**
```
Comment (vikram919) [inline demo-apps/cf-helloworld-server/src/main/java/org/eclipse/californium/examples/MulticastTestServer.java]: new line here

File: demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+			System.out.println("No response received.");
+		}
+
+		client.setURI("coap://" + CoAP.MULTICAST_IPV4.getHostAddress() + "/helloWorld");
+		Request multicastRequest = Request.newGet();
+		multicastRequest.setType(Type.NON);
+		// sends a multicast request
+		client.advanced(handler, multicastRequest);
+		while (handler.waitOn(2000))
+			;
```
Comment (vikram919) [inline demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java]: move semicolon to above line

File: demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	private static final MultiCoapHandler handler = new MultiCoapHandler();
+
+	private static class MultiCoapHandler implements CoapHandler {
```
Comment (vikram919) [inline demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java]: using synchronized onWait() and on() methods looks complicated, a CountDownLatch or a sleep thread of 2000ms can be simple for a demonstration

**Comment by boaks** [comment]
Just to say: **test the function!**

**Comment by vikram919** [comment]
> Just to say: test the function!
I will do that, but these are which caught my eye at first place

File: demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java (hunk 6 above)
Comment (boaks) [reply demo-apps/cf-helloworld-client/src/main/java/org/eclipse/californium/examples/MulticastTestClient.java]: :-).

The idea is to wait until more then 2s no new response arrives. 
synchronize is not complex, especially, if the other primitives just have other functionality.

**Pull request merged**